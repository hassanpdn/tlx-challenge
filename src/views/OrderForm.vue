<template>
      <base-card class="order-form">
            <number-control class="number-input" v-model="order.price" icon="usdCoin" label="Price:" :step="5" :min="0"></number-control>
            <number-control class="number-input" v-model="order.amount" icon="angleRight" label="Quantity:" :step="10" :min="0"></number-control>
            <div class="side-selection">
                  <btn icon="arrowDown" @click="insertOrder('sell')" class="sell" variant="danger" title="sell"/>
                  <btn icon="arrowUp" @click="insertOrder('buy')" class="buy" variant="success" title="buy"/>
            </div>
      </base-card>
</template>

<script>
      //Components
      import BaseCard from '@/components/base/BaseCard';
      import NumberControl from '@/components/base/BaseNumberControlInput';
      import Btn from '@/components/base/BaseButton'

      //Services
      import {insertNewOrder} from '@/services/orders.service'

      export default {
            name: 'order-form',
            components: {BaseCard, NumberControl, Btn},
            data(){
                  return {
                        order: {
                              price: 0,
                              amount: 0,
                              side: ''
                        },
                        selectedOrder: {},
                        insertedOrder: {}
                  }
            },
            methods: {
                  setOrder(src) {
                        const res = {};
                        Object.keys(this.order)
                              .forEach(k => res[k] = (src[k] ?? this.order[k]));
                        this.order = Object.assign({}, this.order, res)
                  },
                  async insertOrder(type){
                        this.order.side = type;
                        const res = await(insertNewOrder(this.order));
                        this.insertedOrder = res.data?.order;
                        this.$root.$emit('notify', {state: res.data.result ? 'success' : 'danger', message: res.data.result || 'Invalid price'})
                  }
            },
            mounted(){
                  this.$root.$on('setOrder', this.setOrder)
            },
            beforeDestroy(){
                  this.$root.$off('setOrder')
            }
      }
</script>

<style>

.order-form {
      margin-top: 58px;
      width: 50%
}

@media screen and (max-width: 48em) {   
      .order-form {
            width: 100%;
      }
}

@media screen and (min-width: 76em) {   
      .order-form {
            width: 30%;
      }
}


.number-input {
      margin-top: 15px;
}

.side-selection {
      display: flex;
      justify-content: space-between;
      margin: 0px auto; 
      text-align: center;
}

.sell, .buy {
      padding-right: 7px;
      padding-left: 7px;
      width: calc(50% - 10px);
}
</style>