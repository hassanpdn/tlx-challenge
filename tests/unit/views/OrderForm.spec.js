

import { mount } from '@vue/test-utils';
import OrderForm from '@/views/OrderForm';


describe('Order Form', () => {
      test("renders component", () => {
            const wrapper = mount(OrderForm);
            expect(wrapper.element).toMatchSnapshot()
      })

      test('should set order', () => {
            const wrapper = mount(OrderForm);
            wrapper.vm.setOrder({price: 20, amount: 2000, side: 'buy'})
            expect(wrapper.vm.order.price).toEqual(20)
      })
      
});
