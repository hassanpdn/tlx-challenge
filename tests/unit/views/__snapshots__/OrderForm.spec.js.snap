// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Order Form renders component 1`] = `
<div
  class="card order-form"
>
  <!---->
   
  <div
    class="card-body"
  >
    <div
      class="control number number-input"
    >
      <button
        class="decrement-button"
      >
        −
      </button>
       
      <button
        class="increment-button"
      >
        +
      </button>
       
      <label
        class="label"
        for="number-control"
      >
        Price:
      </label>
       
      <div
        class="input-icons"
      >
        <svg
          class="icon icon"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M209.2 233.4l-108-31.6C88.7 198.2 80 186.5 80 173.5c0-16.3 13.2-29.5 29.5-29.5h66.3c12.2 0 24.2 3.7 34.2 10.5 6.1 4.1 14.3 3.1 19.5-2l34.8-34c7.1-6.9 6.1-18.4-1.8-24.5C238 74.8 207.4 64.1 176 64V16c0-8.8-7.2-16-16-16h-32c-8.8 0-16 7.2-16 16v48h-2.5C45.8 64-5.4 118.7.5 183.6c4.2 46.1 39.4 83.6 83.8 96.6l102.5 30c12.5 3.7 21.2 15.3 21.2 28.3 0 16.3-13.2 29.5-29.5 29.5h-66.3C100 368 88 364.3 78 357.5c-6.1-4.1-14.3-3.1-19.5 2l-34.8 34c-7.1 6.9-6.1 18.4 1.8 24.5 24.5 19.2 55.1 29.9 86.5 30v48c0 8.8 7.2 16 16 16h32c8.8 0 16-7.2 16-16v-48.2c46.6-.9 90.3-28.6 105.7-72.7 21.5-61.6-14.6-124.8-72.5-141.7z"
            fill="#000"
            opacity="1"
          />
        </svg>
         
        <input
          class="input-field"
          id="number-control"
          max="Infinity"
          min="0"
          step="5"
          type="number"
        />
      </div>
    </div>
     
    <div
      class="control number number-input"
    >
      <button
        class="decrement-button"
      >
        −
      </button>
       
      <button
        class="increment-button"
      >
        +
      </button>
       
      <label
        class="label"
        for="number-control"
      >
        Quantity:
      </label>
       
      <div
        class="input-icons"
      >
        <svg
          class="icon icon"
          viewBox="0 0 256 512"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
            fill="#000"
            opacity="1"
          />
        </svg>
         
        <input
          class="input-field"
          id="number-control"
          max="Infinity"
          min="0"
          step="10"
          type="number"
        />
      </div>
    </div>
     
    <div
      class="side-selection"
    >
      <button
        class="btn sell danger"
      >
        <span>
          sell
        </span>
         
        <svg
          class="icon icon is-button"
          viewBox="0 0 32 32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M26.29 20.29 18 28.59V0h-2v28.59l-8.29-8.3-1.42 1.42 10 10a1 1 0 0 0 1.41 0l10-10z"
            fill="#fff"
            opacity="1"
          />
        </svg>
      </button>
       
      <button
        class="btn buy success"
      >
        <span>
          buy
        </span>
         
        <svg
          class="icon icon is-button"
          viewBox="0 0 32 32"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M26.71,10.29l-10-10a1,1,0,0,0-1.41,0l-10,10,1.41,1.41L15,3.41V32h2V3.41l8.29,8.29Z"
            fill="#fff"
            opacity="1"
          />
        </svg>
      </button>
    </div>
  </div>
</div>
`;
